docs/proposals/auth.md

```diff
@@ -0,0 +1,150 @@
+Heapster Authentication and Authorization
+=========================================
+
+Current State
+-------------
+
+Heapster authentication is based on checking client certificates against a CA.
+Authorization is either just a static list of names, or non-existant.  In
+either case, anyone authenticated is authorized to perform any operation.  With
+push metrics, the proposed plan was to allow adding a second CA, to allow
+separating authorization for model and Oldtimer queries from authorization to
+push metrics.  However, this is suboptimal, since it also separates
+authentication.
+
+Requirements
+------------
+
+Heapster should be able to separate authentication from authorization, and
+the authorization interface should be flexible enough to support authorization
+implementations that are able to make the following distinctions:
+
+- containing queries to a particular namespace: we should be able to only allow
```

Comment:
Number 3 sounds like "yes" to my question about multi-tenancy.

Say you add authorization by namespace.
Then, the next thing that happens is someone sends lots of high-cpu queries and makes the system slow for everyone else.  Or sends to many metrics, and  uses up all the storage.  IF heapster was not designed to be multitenant from the start, and I don't think it was, this is going to be whack-a-mole for a long time. 

The alternative is to run a heapster per namespace.  This solves both authz and qos problems.  If you have a free tier where you can't afford to run heapster per namespace, you can use a shared instance, but restrict access to it.